### ISSUE TITLE
Hasted shopkeepers will block the door without a reason when door is in a corner and player stays close

### ISSUE THREAD

**Issue by ars3niy** [NetHack/NetHack#1267]
It happens to me somewhat regularly that while I am in a shop and do not have debt nor invisibility, the shopkeeper is blocking the door for dozens, or sometimes hundreds, of turns. Turns 3156-3344 of the attached ttyrec is one such example.

Eventually they will move without me doing anything else than moving back and forth.

I do not have a core dump at the moment since I haven't played locally, but I may in the future.

[2024-08-05.08_11_23.ttyrec.gz](https://github.com/user-attachments/files/16570891/2024-08-05.08_11_23.ttyrec.gz)

**Comment by ars3niy** [comment]
Managed to reproduce this locally under GDB. It happens when the door is in a corner and the shopkeeper has generated with a potion of speed (which they will drink when approached) while the player has no speed and keeps hanging out next to the shopkeeper. In my current local game I can directly see that the shopkeeper is fast, and similarly, in the two ttyrecs I have (one above and also another one) there was each time a chugging sound when I was nearing the shop in question.

Now, a hasted shopkeeper has speed of 24, which means they will always move twice each turn. Then as soon as it occurs somehow that the shopkeeper is blocking the door and player is next to shopkeeper while having no bill nor debt, the following infinite loop will occur, courtesy of shk_move and move_special:
* player on their turn performs an action which keeps them next to shopkeeper;
* shopkeeper gets to move twice;
* on the first move satdoor is true and the shopkeeper will choose one of the couple possible squares to move to at random (appr = 0), unblocking the door;
* on the second move satdoor is false and so this time of all possible squares to move to the shopkeeper will chose the one that's closest to the door-blocking location (appr = 1) which (as long as no other monster moved there) will be the door-blocking location itself;
* at no point will the shopkeeper not be lined up with the player because any locations next to the door-blocking square are lined up so shk_move will never hit early return where it checks for !onlineu(omx, omy) (since, by assumption, the player will always be in one of those locations);
* we're back to where we started.

The loop can be broken by player moving further away while still lined up with the door-blocking square where the shopkeeper is at, because then on the first move the shopkeeper can move to where it's no longer lined up with the player, and then on the second move shk_move will hit that early return. Similar thing happens if the door is not in a corner: out of 5 locations inside the shop next to door-blocking square, 4 are such that if the player is there, shopkeeper can on the first move go to where they're not lined up with the player and hit early return on the second move.

The fact that it took me a month of playing and 2 hours in GDB to figure out how to reliably get out of a shop suggests that this logic needs to be improved.
